# <center>Chapter_02</center>

```python
# 各种需要用到的函数

"""

input()                    函数接受一个标准输入数据，返回为字符串str类型
int()                      将输入转化为整数
math.factorial()           计算阶乘
numpy.cumprod()            计算累积乘积
numpy.inner()              计算行向量的内积，函数输入为连个列向量时得到的结果为张量积
numpy.linalg.inv()         计算方阵的逆
numpy.linspace()           在指定的间隔内，返回固定步长的数组
numpy.math.factorial()     计算阶乘
numpy.random.seed()        固定随机数发生器种子
numpy.random.uniform()     产生满足连续均匀分布的随机数
numpy.sum()                求和
scipy.special.factorial()  计算阶乘
seaborn.heatmap()          绘制热图

"""

```

## 2.1 算数乘除：先乘除，后加减，括号内先算
### 2.1.1 乘法
**乘法**算式等号左端是**被乘数**和**乘数**，右端是**乘积**。
几个常用的乘法符号：
- ${\times}$用于数字相乘，一般不用于变量相乘。
    - 在线性代数中${\times}$表示**叉乘**
- 两变量a,b相乘可以写成${ab}$，这种记法被称为**隐含乘法**，${ab}$也可写为${a \cdot b}$  
乘法运算时要注意单位。有一些数值本身**无单位**，比如个数，z分数（**标准分数**）。

In [7]:
# 计算两个数的乘法

num1 = 2
num2 = 3

# num1 times num2 equals prod
prod = num1*num2

# display the computation
print('The product of {0} and {1} is {2}'.format(num1, num2, prod))

The product of 2 and 3 is 6


### 2.1.2 阶乘
某个正整数的**阶乘**是所有小于等于该数的正整数的积。例如5的阶乘记为${5!}$：  
$${5!=5 \times 4 \times 3 \times 2 \times 1}$$
**特别的，0!=1**

In [3]:
# 计算一个整数的阶乘
number = int(input("Please enter a integer: "))
factorial = 1

if number < 0:
    print("The integer should be positive.")
elif number == 0:
    print("The integer of 0 is 1")
else:
    for i in range(1, number + 1):
        factorial *= i
    print("The factorial of {0} is {1}".format(number, factorial))

Please enter a integer:  5


The factorial of 5 is 120


### 2.1.3 累计乘积
对于一组数字，**累计乘积**也叫**累积乘积**，类似于累计求和，，它得到的结果不仅仅是一个乘积，二十从左向右每乘一个数值得到的分步结果。比如，自然数1到10的累计求积结果为：
$${1, 2, 6, 24, 120, 720, 5040, 40320, 362880, 3628800}$$  

In [7]:
# 计算一组数的累积乘积

import numpy as np

ten_num = np.linspace(1, 10, 10)
print(ten_num)

products = np.cumprod(ten_num)
np.set_printoptions(suppress=True)  # 禁用科学计数法
print(products)

"""使用numpy.arange()生成1-10"""
ten_num = np.arange(1, 11)
print(ten_num)

[ 1.  2.  3.  4.  5.  6.  7.  8.  9. 10.]
[      1.       2.       6.      24.     120.     720.    5040.   40320.
  362880. 3628800.]
[ 1  2  3  4  5  6  7  8  9 10]


### 2.1.4 除法
**除法**是**乘法的逆运算**。**被除数除以除数得到商**。  
除法运算有时候可以**除尽**；有时候会得到**余数**，例如7除2余1。  
除法的结果一般用分数或小数来表达。

In [8]:
# 两个数的除法

num1 = 6
num2 = 3

# num1 divided by num2 equals division
division = num1/num2

# display the computation
print('The division of {0} over {1} is {2}'.format(num1, num2, division))

The division of 6 over 3 is 2.0


In [9]:
# 两个数除法（取余）

num1 = 7
num2 = 2

# 'remainder' is the remainder of num1 divided by num2
remainder = num1%num2

# display the computation
print('The remainder of {0} over {1} is {2}'.format(num1, num2, remainder))

The remainder of 7 over 2 is 1


### 2.1.5 分数
最常见的分数是**普通分数**，由**分母**和**分子**组成，分隔二者的是**分数线**或正斜杠${/}$。  
非零整数a的**倒数**是$\frac{1}{a}$，分数$\frac{b}{a}$的倒数是$\frac{a}{b}$。

### 2.1.6 英文对照
|数学表达|英文表达|
|:---:|:---|
|${2 \times b =6}$|Two times three equals six<br>Two multiplied by three equals six.<br>Two cross three equals six.<br>The product of two and three is six.<br>If you multiply two by three, you get six.|
|${a \cdot b=c}$|*a* times *b* equals *c*.<br>*a* multiplied by *b* equals *c*.<br>*a* dot *b* equals *c*.<br>The product of *a* and *b* is *c*.|
|${6 \div 3=2}$|Six divided by three equals two.<br>If you divide six by three, you get two.|
|${7 \div 2=3 R 1}$|Seven over two is three and remainder is one.<br>Seven divided by two equals three with a remainder of one.<br>Seven divided by two equals three and the remainder is one.|
|$\frac{1}{2}$, ${1/2}$|One half<br>A half<br>One over two|
|${1:2}$|One to two|
|$-\frac{3}{2}$|Minus three-halves<br>Negative three-halves|
|$\frac{1}{3}$, ${1/3}$|One over three<br>One third|
|$\frac{1}{4}$, ${1/4}$|One over four<br>One fourth<br>One quater<br>One divided by four|
|${1}\frac{1}{4}$|One and one fourth|
|$\frac{3}{5}$|Three fifths|
|$\frac{1}{n}$|One over n|
|$\frac{a}{b}$|a over b<br>a divided by b<br>The ratio of a to b<br>The numerator is *a* while the denominator is *b*.|  
<center>表1. 数学英文对照表</center>

|英文名词|汉语名词|
|:---|:---|
|**multiplication**|乘法|
|**multiplicand**|被乘数|
|**multiplier**|乘数|
|**product**|乘积|
|**times, multiplied by**|乘|
|**multiplication table, times table**|乘法表|
|**cross product, vector product**|叉乘|
|**implied multiplication**|隐含乘法|
|**decimal point**|小数点|
|**scalar product**|标量积|
|**unitless**|无单位|
|**standard score**|标准分数|
|**fractorial**|阶乘|
|**cumulative product**|累计乘积|
|**division**|除法|
|**dividend**|被除数|
|**over, divided by**|除以|
|**divisor, denominator**|除数|
|**quotient**|商|
|**divisible**|除尽|
|**remainder**|余数|
|**fraction**|分数|
|**decimal**|小数|
|**denominator**|分母|
|**numerator**|分子|
|**fraction bar**|分数线|
|**reciprocal**|倒数|  
<center>表2. 中英文对照表</center>

## 2.2 向量乘法：标量乘法、向量内积、逐项积

In [12]:
# 向量和矩阵的标量乘法



In [2]:
# 向量内积

import numpy as np




In [1]:
# 逐项积

import numpy as np


## 2.3 矩阵乘法：最重要的线性代数运算规则
### 2.3.1 矩阵乘法规则
${A}$和${B}$两个矩阵相乘的前提是**矩阵${A}$的列数和矩阵${B}$的行数相同**。  
${A}$和${B}$的乘积一般写作${AB}$。  
在Numpy中，两个矩阵相乘的运算符为@，${AB}$也记作${A@B}$：
$${{C}_{m \times n}={A}_{m \times p}{B}_{p \times n}={A}_{m \times p}@{B}_{p \times n}}$$  
矩阵乘法不满足交换律，即通常情况下：
$${{A}_{m \times p}{B}_{p \times n} \neq {B}_{p \times n}{A}_{m \times p}}$$

### 2.3.2 两个${2 \times 2}$矩阵相乘
下式中，矩阵${A}$和${B}$相乘结果为矩阵${C}$：
$${{C}={AB}=\begin{bmatrix}1 & 2 \\ 3 & 4\end{bmatrix}\begin{bmatrix}4 & 2 \\ 3 & 1\end{bmatrix}=\begin{bmatrix}1 & 2 \\ 3 & 4\end{bmatrix}@\begin{bmatrix}4 & 2 \\ 3 & 1\end{bmatrix}}$$  
矩阵${A}$的第一行元素和矩阵${B}$第一列对应元素进行相乘，再相加，结果得到矩阵${C}$的第一行、第一列元素${{c}_{1,1}}$。  
矩阵${A}$的第一行元素和矩阵${B}$第二列对应元素进行相乘，再相加，结果得到矩阵${C}$的第一行、第一列元素${{c}_{1,2}}$。
${\cdots}$  
总体来说，${A}$和${B}$的乘积${C}$第i行第j列元素${{c}_{i,j}}$等于矩阵${A}$的第i行的元素与矩阵${B}$第j列对应元素乘积再求和。
![图8.矩阵乘法规则，两个${2 \times 2}$矩阵相乘为例](images/8.png)
<center>图8.矩阵乘法规则，两个${2 \times 2}$矩阵相乘为例</center>

In [2]:
# 矩阵乘法

import numpy as np

array_1 = np.array([[1, 2, 5], [6, 4, 7], [3, 8, 9]])
array_2 = np.array([[4, 5, 7], [9, 8, 3], [1, 5, 6]])

prod_array = array_1 @ array_2
print(prod_array)

[[ 27  46  43]
 [ 67  97  96]
 [ 93 124  99]]


### 2.3.3 矩阵乘法形态
图9中列出了常见的矩阵乘法形态，每种形态对应一类线性代数问题。
![图9.矩阵乘法形态多样性](images/9.png)
<center>图9.矩阵乘法形态多样性</center>

### 2.3.4 两个向量相乘
1. 向量${a}$和${b}$为等长列向量，${a}$转置(${a}^{T}$)乘${b}$为标量，等价于${a}$和${b}$的标量积：
$${{a}^{T}b={a \cdot b}}$$  
例如：
$${{a}^{T}b=\begin{bmatrix}1 \\ 2 \\ 3\end{bmatrix}^{T}_{3 \times 1}@\begin{bmatrix}4 \\ 3 \\ 2\end{bmatrix}_{3 \times 1}=\begin{bmatrix}1 & 2 & 3\end{bmatrix}_{1 \times 3}@\begin{bmatrix}4 \\ 3 \\ 2\end{bmatrix}_{3 \times 1}=\begin{bmatrix}1 & 2 & 3\end{bmatrix} \cdot \begin{bmatrix}4 & 3 & 2\end{bmatrix}=16}$$  

2. 列向量${a}$乘${b}$转置(${b}^{T}$)，乘积结果${ab^{T}}$为矩阵：
$${{a}b^{T}=\begin{bmatrix}1 \\ 2 \\ 3\end{bmatrix}_{3 \times 1}@\begin{bmatrix}4 \\ 3 \\ 2\end{bmatrix}^{T}_{3 \times 1}=\begin{bmatrix}1 \\ 2 \\ 3\end{bmatrix}_{3 \times 1}@\begin{bmatrix}4 & 3 & 2\end{bmatrix}_{1 \times 3}=\begin{bmatrix}4 & 3 & 2 \\ 8 & 6 & 4 \\ 12 & 9 & 6\end{bmatrix}}$$  
$${{a}b^{T}=\begin{bmatrix}1 \\ 2\end{bmatrix}_{2 \times 1}@\begin{bmatrix}4 \\ 3 \\ 2\end{bmatrix}^{T}_{3 \times 1}=\begin{bmatrix}1 \\ 2\end{bmatrix}_{2 \times 1}@\begin{bmatrix}4 & 3 & 2\end{bmatrix}_{3 \times 1}=\begin{bmatrix}4 & 3 & 2 \\ 8 & 6 & 4\end{bmatrix}}$$  

## 2.4 矩阵乘法的第一视角
